main.go

```diff
@@ -79,23 +79,20 @@ func main() {
 
 	sources := source.NewMultiSource(source.LookupMultiple(cfg.Sources...)...)
 
-	googleProvider, err := dnsprovider.NewGoogleProvider(cfg.GoogleProject, cfg.DryRun)
-	if err != nil {
-		log.Fatal(err)
-	}
+	dnsprovider.Register("google", dnsprovider.NewGoogleProvider(cfg.GoogleProject, cfg.DryRun))
+	dnsprovider.Register("aws", dnsprovider.NewAWSProvider(cfg.DryRun))
```

Comment:
yes, sounds right. However, you would just move the following lines into a method in the `dnsprovider` package (I wouldn't consider this a change).

The registration part here is more annoying. It needs provider specific configuration passed  in, e.g. the google project. So we still need it this way I believe. We could pass arbitrary config to your proposed `Initialize` method and then just forward to the picked provider, but I'm not sure if that would be any better.